Ekubo wETH-ETH liquidity pool.ipynb

```diff
@@ -0,0 +1,1405 @@
+    "            il = impermanent_loss(price_ratio)\n",
+    "            impermanent_loss_total += il\n",
+    "            \n",
+    "            # Update liquidity position by adding earned fees\n",
+    "            current_liquidity += fees_earned\n",
+    "            fees_earned = 0  # Reset fees after rebalancing\n",
+    "            \n",
+    "            # Reset initial price to the new price after rebalancing\n",
+    "            initial_price = current_price\n",
+    "            print('i')\n",
+    "    \n",
+    "    # Final calculation of returns\n",
+    "    net_earnings = fees_earned - impermanent_loss_total\n",
+    "    \n",
+    "    return net_earnings, fees_earned, impermanent_loss_total\n",
+    "\n",
+    "# Run the simulation with the updated logic and generate the optimization results\n",
+    "optimization_results = []\n",
+    "liquidity_range_options = np.arange(0.001, 0.021, 0.005)  # Liquidity add range from 5% to 20%\n",
+    "rebalance_range_options = np.arange(0.001, 0.021, 0.005)  # Rebalance range from 5% to 20%\n",
+    "\n",
+    "for liquidity_range in liquidity_range_options:\n",
+    "    for rebalance_range in rebalance_range_options:\n",
+    "        net_earnings, fees_earned, impermanent_loss_total = run_simulation(liquidity_range, rebalance_range)\n",
+    "        optimization_results.append({\n",
+    "            \"Liquidity Range (%)\": liquidity_range * 100,\n",
+    "            \"Rebalance Range (%)\": rebalance_range * 100,\n",
+    "            \"Net Earnings (ETH)\": net_earnings,\n",
+    "            \"Total Fees Earned (ETH)\": fees_earned,\n",
+    "            \"Impermanent Loss (ETH)\": impermanent_loss_total\n",
+    "        })\n",
+    "\n",
+    "# Convert optimization results to a DataFrame for analysis\n",
+    "optimization_df = pd.DataFrame(optimization_results)\n",
+    "\n",
+    "# Sort by net earnings to find the best performing combination\n",
+    "optimization_df_sorted = optimization_df.sort_values(by=\"Net Earnings (ETH)\", ascending=False)\n",
+    "\n",
+    "# Display the top 5 performing combinations\n",
+    "print(optimization_df_sorted.head())\n"
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": null,
+   "id": "1deb2c2a",
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "# Verify prices\n",
+    "\n",
+    "# Define the ticker symbols for ETH and wETH\n",
+    "eth_symbol = 'ETH-USD'\n",
+    "weth_symbol = 'WETH-USD'\n",
+    "\n",
+    "# Fetch historical data for both tokens over a given period (e.g., 1 month)\n",
+    "eth_data = yf.Ticker(eth_symbol).history(period=\"1mo\")\n",
+    "weth_data = yf.Ticker(weth_symbol).history(period=\"1mo\")\n",
+    "\n",
+    "# Align both dataframes by the same date index (merge them on the 'Date' column)\n",
+    "data = pd.DataFrame({\n",
+    "    'ETH Price': eth_data['Close'],\n",
+    "    'wETH Price': weth_data['Close']\n",
```

Comment:
Tests are unfinished. Before finishing tests, writing strategy doesnt make sense as u will never know if u r doing right or wrong math.

Reply:
Above strategy code will be reviewed after finishing tests. Look at pytest and configure tests for the project.